lib/spack/spack/repo.py

```diff
@@ -1111,6 +1111,10 @@ def _get_pkg_module(self, pkg_name):
             module.__package__ = self.full_namespace
             module.__loader__ = self
             self._modules[pkg_name] = module
+            if fullname in sys.modules:
+                # In this case we have a new Repo object, but a prior Repo
+                # has loaded the module (this only occurs during testing)
+                del sys.modules[fullname]
```

Comment:
If your PRs pass unit tests, fix the associated issues, and don't do this, then I would say none (from an experimental if-it-works-it-works viewpoint).

Otherwise:

I originally added this because there are some tests that modify the Package class that is imported, and the `mutable_mock_repo` fixture was intended for these cases to ensure that these modifications did not pollute other tests (i.e. a test using `mutable_mock_repo` imports and modifies a package `x` and a later test uses `x` but expects that none of the other modifications are present). This pollution would theoretically occur when switching to/from a test using `mutable_mock_repo`. 

(kind of a side note but relevant for doing spawn-based subprocesses on Linux) Swapping between `mutable_mock_repo` and `mock_packages` can also cause problems for pickling: conflicts between `get_package_class` and the import method provided by `Repo` can lead to a case where the Package class comes from a module that is not in `sys.modules`.

Recently I became suspicious that this is not sufficient (Python import docs imply that deleting from `sys.modules` doesn't necessarily have the effect one would desire, since it doesn't modify the underlying structures used to track imports in the Python implementation): I don't know if there is any way to truly "unload" a module, so I was planning to perform all such tests in a subprocess. As part of this PR I started removing uses of `mutable_mock_repo` where they weren't required in order to get a better sense of how much work that would be.

Reply:
Thanks for the explanation @scheibelp 

> If your PRs pass unit tests, fix the associated issues, and don't do this, then I would say none (from an experimental if-it-works-it-works viewpoint)

I think this PR is an improvement over what we have in `develop` and points out problematic aspects of the current implementation. If you agree, now that #20005 has been worked around, I think we should proceed with less pressure to get this PR, #23290 and #23269 merged.